publisher/elasticsearch.go

```diff
@@ -0,0 +1,275 @@
+	client := HTTPClientWithRetry()
+	var (
+		bulkdata []byte
+		err      error
+	)
+	const bulkStart = "{\"index\":{}}\n"
+	const bulkEnd = "\n"
+
+	for _, doc := range data {
+		byteData, err := json.Marshal(doc)
+		if err != nil {
+			log.Errorf("error[%v] unable to marshal data", err)
+		} else {
+			bulkdata = append(bulkdata, bulkStart...)
+			bulkdata = append(bulkdata, byteData...)
+			bulkdata = append(bulkdata, bulkEnd...)
+		}
+	}
+
+	bulkdata = append(bulkdata, bulkEnd...)
+
+	reqURL := fmt.Sprintf("%s_bulk/", es.getURL())
+	request, err := retryhttp.NewRequest(
+		http.MethodPost,
+		reqURL,
+		bytes.NewReader(bulkdata),
+	)
+	if err != nil {
+		log.Errorf("error[%v] unable to create new request", err)
+		return err
+	}
+	log.Debugf("request: %+v, body: %+v, ", request.Request, request.Body)
+
+	request.Header.Set("Content-Type", "application/json")
+	if es.Username != "" && es.Password != "" {
+		request.SetBasicAuth(es.Username, es.Password)
+	}
+
+	timeout, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	reqWithTimeout := request.WithContext(timeout)
+	if es.Protocol == "https" {
+		client.HTTPClient.Transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+			Proxy: http.ProxyFromEnvironment,
+		}
+	}
+
+	response, err := client.Do(reqWithTimeout)
+	if err != nil {
+		log.Errorf("error[%v] request failed with error", err)
+		return err
+	}
+	defer response.Body.Close()
+
+	respBody, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		log.Errorf("error[%v] failed to read response body", err)
+	}
+	log.Debugf("respBody after writing to elastisearch\n%s", string(respBody))
+
+	var bulkResp BulkResponse
+	err = json.Unmarshal(respBody, &bulkResp)
+	if err != nil {
+		log.Errorf("error[%v] failed to Unmarshal response body", err)
+		return errors.New(response.Status)
+	}
+	log.Debugf("error[%v] bulkResp", bulkResp)
+
+	bulkrespErr := []error{}
+	if bulkResp.Errors {
+		for _, item := range bulkResp.Items {
+			idoc := item["index"]
+			if idoc.Status > 299 {
+				bulkrespErr = append(
+					bulkrespErr,
+					fmt.Errorf("doc error: %s", idoc.Error.Reason),
+				)
+			}
+		}
+		log.Errorf("error[%v] response errors", bulkrespErr)
+		return nil
+	}
+
+	if response.StatusCode == http.StatusOK || response.StatusCode == http.StatusCreated {
+		log.Infof("successfully sent doc to %s - %d", es.getURL(), response.StatusCode)
+		return nil
+	}
+	log.Errorf("[elasticsearch] failed to send doc to %s - %d", es.getURL(), response.StatusCode)
+	var esError ErrorResponse
+	err = json.Unmarshal(respBody, &esError)
+	if err != nil {
+		log.Errorf("error[%v] failed to Unmarshal response body, %s", err)
+		return errors.New(response.Status)
+	}
+	log.Errorf("error[%v]", esError)
+	return fmt.Errorf("status code: %d, error: %s", esError.Status, esError.Error.Reason)
+}
+
+// getURL returns url to elasticendpoint for writing data
```

Comment:
Check the elastic beats code
https://github.com/elastic/beats/tree/main/libbeat/outputs

There may be some cases like index does not exist or generating index name based on some fields on doc.
Same for Kafkarest

Reply:
https://github.com/snappyflow/beats/tree/master/libbeat/outputs